[1단계 - 자동차경주] 황의찬 미션 제출합니다.

pine14 opened the pull request.

Commit:
build : UTF-8 인코딩 설정

build.gradle에 UTF-8 인코딩 설정을 추가했습니다.

Commit:
docs : 문자열 계산기 요구사항 문서 추가

문자열 계산기의 요구사항 첫 번째 :
 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환

Commit:
test : 쉼표 구분자, 쉼표 또는 콜론 구분자에 대한 테스트 작성

실제 코드를 작성하기 전에 첫 번째 요구사항인
쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환
에 대한 테스트를 먼저 작성하였습니다.

Commit:
feat : 구분자를 기준으로 분리한 각 숫자의 합을 반환하는 기능 구현

String input을 정규표현식을 이용해서 자르고 숫자로 변환한 후 합계를 계산하는 기능을 구현했습니다.

Commit:
docs : 커스텀 구분자 요구사항 추가

앞의 기본 구분자(쉼표, 콜론)외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열
앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다.

Commit:
test : 커스텀 구분자 테스트 케이스 작성

Commit:
feat : 커스텀 구분자를 이용한 문자열 합 구현

Matcher, Pattern을 이용해서 커스텀 구분자를 이용한 문자열의 합을 구현했습니다.

Commit:
test : null 또는 빈문자, 숫자하나의 입력 테스트 작성

Clean Code 9장 단위 테스트에 따르면 테스트 당 개념 하나를 테스트하는 것이 좋다고 나와있습니다.
따라서 splitAndSum_null_또는_빈문자() 테스트를
splitAndSum_null() 과 splitAndSum_빈문자() 로 나누었습니다.

Commit:
docs : 예외처리 요구사항 정의

문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw합니다.

Commit:
숫자 이외의 값 또는 음수를 전달할 때의 테스트 작성

Commit:
음수를 입력하면 RuntimeException이 나오도록 검증하는 기능 구현

RuntimeException 예외를 throw하도록 처리했습니다.

Commit:
test : 음수 입력 시와 숫자 이외의 값 테스트 분리

Clean Code 테스트 당 개념 하나 에 따라 두 개의 테스트 메서드로 분리했습니다.

Commit:
refactor : 매직 넘버 피하기

매직 넘버 대신 상수를 사용해서 가독성을 높였습니다.

Commit:
refactor : 문자열의 합계 구하는 부분을 Stream으로 변경

기존의 반복문에서 stream을 사용하는 코드로 변경했습니다.

Commit:
refactor : indent를 줄이기 위해 getSum 리팩토링

depth를 1단계로 줄이기 위해 반복을 돌면서 음수를 검증하는 방식으로 바꾸었습니다.

Commit:
docs : 자동차 경주 게임 1단계 요구사항 정리

각 자동차에 이름을 부여할 수 있다. 전진하는 자동차를 출력할 때 자동차 이름을 같이 출력한다.
자동차 이름은 쉼표(,)를 기준으로 구분하며 이름은 5자 이하만 가능하다.

Commit:
test : 자동차 생성 테스트 작성

자동차 생성 테스트를 먼저 작성한 후에 Car 클래스를 만들었습니다.

Commit:
refactor : BeforeEach를 사용해서 CarTest 리팩토링

매 테스트마다 Car를 생성하도록 BeforeEach를 구현했습니다.

Commit:
test : 자동차의 초기 위치는 0이어야 하는 테스트 구현

Commit:
feat : 자동차의 이름을 입력받는 view 구현

Scanner를 사용해 자동차와 시도 횟수를 입력받았습니다.

Commit:
test : 자동차 이름 생성 성공 테스트

자동차의 이름을 생성하는 것에 성공했을 때의 테스트를 작성했습니다

Commit:
docs : 자동차 이름 예외처리 추가

이름은 5자 이하만 가능합니다.
자동차 이름은 비어있을 수 없습니다.
자동차 이름은 중복이 불가능합니다.

Commit:
docs : 자동차 이름 예외처리 변경

쉼표로 나누기 전과 나누었을 때의 검증을 구분했습니다.

Commit:
test : 자동차 이름 검증 테스트 작성

테스트를 위해 일급 컬렉션 Cars를 구현했습니다.
자동차 이름은 null일 수 없으므로 그에 맞는 테스트를 작성했습니다.

Commit:
refactor : 자동차 이름 검증 테스트 이름 변경

null을 체크하는 클래스이므로 해당 기능에 맞는 이름으로 변경했습니다.

Commit:
test : 자동차 이름 검증 테스트 구현 완료

자동차 이름을 검증하는 테스트 코드를 작성한 후
NameValidator 클래스에 구현했습니다.

Commit:
feat : 검증 로직 추가

Car의 생성자와 Cars의 add 메서드에서 이름에 대한 검증을 수행하도록 처리했습니다.

Commit:
docs : 자동차 경주 게임 요구사항 추가

사용자는 몇 번의 이동을 할 것인지를 입력할 수 있어야 합니다.

Commit:
test : 사용자로부터 숫자를 입력받는 테스트 구현

시도횟수를 정상 입력했을 때의 테스트를 구현했습니다.

Commit:
feat : 사용자로부터 숫자를 입력받아 정수로 변환

Commit:
docs : 사용자의 입력 횟수 예외처리 요구사항

1 미만의 수를 입력했을 경우와
숫자를 입력하지 않았을 경우 예외를 발생해야 합니다.

Commit:
feat : 사용자 입력 횟수 예외처리 구현

요구사항에 맞게 예외처리를 진행했습니다.

Commit:
docs : 전진하는 조건 요구사항 추가

주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
전진하는 조건은 0에서 9 사이에서 random 값을 구한 후 random 값이 4 이상일 경우 전진하고, 3 이하의 값이면 멈춘다.

Commit:
feat : 사용자의 시도 횟수 검증하는 코드 사용

NumberGenerate에서 validateNumber를 사용해 검증합니다.

Commit:
test : 자동차 전진 테스트 작성

랜덤값이 4 이상일 경우 전진합니다.
자동차가 전진하지 못하면 위치는 그대로입니다.
자동차가 전진한다면 위치가 1 증가해야 합니다.

Commit:
feat : 자동차 경주 게임 Controller 구현

ready() 메서드에서 자동차 경주 게임에 관한 일을 처리합니다.
OutputData에서 출력 view를 처리합니다.

Commit:
feat : 랜덤 값을 생성하는 클래스 RandomNumberGenerator 클래스 생성

랜덤 값을 생성하는 부분을 클래스로 분리했습니다.

Commit:
docs : 자동차 경주 게임을 완료한 후 누가 우승했는지 알려줍니다

Commit:
test : 가장 멀리 이동한 자동차를 찾는 테스트 작성

자동차를 3번 앞으로 전진시킨 후 가장 멀리 이동한 자동차의 위치를 찾는 테스트를 구현했습니다.

Commit:
feat : 자동차 게임 우승자를 찾는 코드 구현

자동차 게임 공동 우승자와 단독 우승자를 찾는 테스트를 만든 후
stream을 이용해 구현했습니다.

Commit:
feat : 최종 우승자 출력

String.join 을 이용해 최종 우승자를 출력합니다.

Commit:
refactor : 검증 테스트 수정

Commit:
refactor : 테스트 이름 변경

Commit:
test : ParameterizedTest 구현

생성자에서 예외가 터져 테스트가 제대로 진행되지 않았던 점을 수정하고
자동차 예외 검증을 한번에 처리할 수 있는 테스트를 ParameterizedTest를 이용해서 구현했습니다.

Commit:
refactor : 불변 리스트 반환 시 List.copyOf 사용

Java9에서 도입된 List.copyOf를 사용해 Immutable List를 반환합니다.

Commit:
refactor : NameValidatorTest AssertJ 사용

위의 테스트와 동일하게 Assertj를 사용하도록 변경했습니다.

Commit:
refactor : 매직 넘버를 피하기

매직넘버의 사용은 코드의 가독성을 어렵게 만드므로 상수를 사용했습니다.

skullkim commented on build.gradle:
```diff
@@ -11,6 +11,8 @@ repositories {
 dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
     testImplementation 'org.junit.jupiter:junit-jupiter:5.8.2'
+    compileJava.options.encoding = 'UTF-8'
```
인코딩을 지정한 이유가 있을까요?🧐

pine14 replied:
https://naver.github.io/hackday-conventions-java/
[캠퍼스 핵데이 Java 코딩 컨벤션]의 
1.1 파일 인코딩은 UTF-8 = 모든 소스, 텍스트 문서 파일의 인코딩은 UTF-8로 통일한다.

을 보고 지정했습니다.

skullkim commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,55 @@
+
+import org.junit.jupiter.api.Test;
+
+class CalculatorTest {
```
모든 테스트 메서드에 throws Exception이 존재하는 이유가 있을까요?

pine14 replied:
(인프런에서 스프링 강의로 유명한) 강사님의 Intellij Live Template을 보고 test를 작성하다 보니 throws Exception 코드가 들어간 것 같습니다.😂

예전에는 생각없이 사용했는데,
```java
@Test
public void test() throws Exception {

}
```
이와 같이 Exception을 던지는 코드를 기본적으로 사용하는 이유가 뭘까요..?

skullkim replied:
흠.... 명확한 이유는 그 강의를 봐야 알 수 있을거 같네요. 제 추측으로는 테스트 로직에서 checked exception을 던지는 코드가 존재하지 않았나 싶네요🧐

pine14 replied:
checked exception 에 대해 말씀해 주셔서 
checked exception과 unchecked exception에 대해 공부했습니다!

테스트 코드에서는 checked exception을 던지지 않으므로 삭제했습니다.

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,61 @@
+
+	private static String[] split(String input) {
+		Matcher matcher = getMatcher(input);
+		if (isCustom(matcher)) { // 커스텀 구분자를 사용한다면
```
주석을 사용하셨네요? 의미있는 주석은 코드를 읽는 대 분명 도움이 되요. 하지만 코드를 수정할 때 주석도 같이 수정해 주지 않는다면 코드에 대해 잘못된 이해를 할 수도 있다는 문제가 있죠. 이 경우 주석을 사용하지 않더라도 isCustom()이라는 메서드 이름을 isCustomDelimiter()로 바꾸면 될거 같아요. 어떻게 생각하시나요🧐?

pine14 replied:
메서드 이름만으로 의도를 나타낼 수 있으므로 좋은 것 같습니다!

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,61 @@
+		Matcher matcher = getMatcher(input);
+		if (isCustom(matcher)) { // 커스텀 구분자를 사용한다면
+			String customDelimiter = matcher.group(DELIMITER_NUMBER);
+			return matcher.group(INPUT_GROUP).split(customDelimiter);
```
저같은 경우 메서드 체인이 발생했을 때 메서드 하나 당 한줄을 차지하는 것을 좋아해요. 아래 예시 처럼요.
```java
matcher.group(INPUT_GROUP)
    .split(customDelimiter);
```
이렇게 하는 가장 큰 이유는 가독성 때문인대요. 두 개 까지는 괜찮은거 같기도 하지만 저는 무조건 한 줄에 하나만 쓰는게 보기 편하더라고요😄. 의찬님은 어떻게 작성하는게 더 편하신가요?

pine14 replied:
Stream을 사용하는 코드 convention에서도 메서드 하나 당 한 줄을 사용했는데,
말씀해주신 대로 사용하는게 더 보기 편한 것 같습니다!

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,61 @@
+	private static boolean isCustom(Matcher matcher) {
+		return matcher.find();
+	}
+
+	private static Matcher getMatcher(String input) {
+		return Pattern.compile(CUSTOM_SPLIT_REGEX).matcher(input);
+	}
```
정규식 일치 여부를 확인할 때 메서드 내에서 Patterm.compile()을 그냥 사용하게 되면 메서드가 호출될 때마다 컴파일이 발생해 성능에 영향을 미쳐요😭. 
아래와 같이 바꾸는게 좋아요!
```suggestion
private static final Pattern COMPILED_CUSTOM_SPLIT_REGEX = Pattern.compile(CUSTOM_SPLIT_REGEX);

private static Matcher getMatcher(String input) {
	return COMPILED_CUSTOM_SPLIT_REGEX.matcher(input);
}
```

pine14 replied:
기존 코드는 getMatcher 메서드가 호출될 때 마다 Pattern.compile 을 호출하네요..😭

https://medium.com/webeveloper/%EC%9E%90%EB%B0%94-%EC%98%AC%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%A0%95%EA%B7%9C%EC%8B%9D-%ED%8C%A8%ED%84%B4-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0-c2479c33ab8b

패턴을 여러 번 사용하는 경우 한 번 컴파일한 후 재사용
* 매번 이 메서드를 호출하는 것보다 효율적입니다.

라고 하네요! 수정 완료했습니다.😄

skullkim replied:
수정한 부분 확인했습니다~. 그런대 상수의 경우 대문자로 표기하는게 컨벤션 아니였나요? pattern이 소문자로 되어있네요...

pine14 replied:
아 그렇네요😅 대문자 컨벤션을 지켜 수정했습니다!

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,61 @@
+	}
+
+	private static void checkInputIsNegative(String input) {
+		if (getParseInt(input) < 0) {
```
매직넘버는 사용하지 말아주세요~. 숫자는 그 자체로 어떤 의미를 가지지 않아요!

pine14 replied:
NEGATIVE_NUMBER_STANDARD 와 같이 명명된 상수로 교체했습니다.

skullkim commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,61 @@
+
+	private static void checkInputIsNegative(String input) {
+		if (getParseInt(input) < 0) {
+			throw new RuntimeException("숫자 이외의 값 또는 음수는 계산할 수 없습니다.");
```
RunTimeException 보다 조금 더 명확한 예외를 사용하는게 어떨까요? RuntimeException만 사용한다면 어떤 이유로 예외가 발생했는지 알 수 없어요...

pine14 replied:
IllegalArgumentException을 던지고 싶었으나 문자열 계산기 요구사항에

**문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw한다**

라는 내용이 있어 RuntimeException을 사용했습니다.

요구사항에 RuntimeException으로 지정된 이유가 있을까요..?

skullkim replied:
RuntimeException 예외를 throw 한다는 요구 사항은 unchecked exception을 throw 하라는 의미로 사용한 것 같네요. IllegalArgumentException도 결국엔 RuntimeException을 throw 하니 사용해도 되지 않을까요?

pine14 replied:
예외에 대한 이해가 부족한 것 같아 **이펙티브 자바 item 72의 표준 예외를 사용하라**를 읽어보았습니다!

`Exception, RuntimeException, Throwable, Error는 직접 재사용하지 말자. 이 예외들은 다른 예외들의 상위 클래스이므로, 즉 여러 성격의 예외들을 포괄하는 클래스이므로 안정적으로 테스트할 수 없다.`

unchecked exception인 IllegalArgumentException을 던지는 것이 맞다고 판단되어 수정했습니다.

skullkim commented on src/test/java/racingcar/model/CarsTest.java:
```diff
@@ -0,0 +1,61 @@
+
+	@Test
+	@DisplayName("가장 멀리 간 위치 찾기")
+	void max_position() {
```
테스트 메서드 이름에 카멜케이스를 사용한 이유가 있을까요?

pine14 replied:
https://google.github.io/styleguide/javaguide.html
문서에 
Underscores may appear in JUnit test method names to separate logical components of the name, with each component written in lowerCamelCase, for example transferMoney_deductsFromSource. There is no One Correct Way to name test methods.

라고 나와있어서 사용했습니다!

skullkim replied:
구글 컨벤션 까지 읽으셨군요😄. 좋아요👏

skullkim commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	@Test
+	void 랜덤값이_4이상이면_전진한다() {
```
테스트는 문서의 목적도 가지고 있으므로 메서드 이름을 한글로 해서 의도를 서술해도 된다고 생각해요~. 하지만 다른 메서드와의 컨벤션은 반드시 일치시키는게 좋아요!. @DisplayName()을 사용하는 방식과 메서드 이름을 한글로 사용하는 방식 중 하나만 상요해 주세요!

pine14 replied:
@DisplayName을 사용하는 방식으로 변경했습니다!👍

skullkim commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,58 @@
+
+	@Test
+	void 랜덤값이_4이상이면_전진한다() {
+		assertThat(car.shouldMove(4)).isTrue();
```
하나의 유닛 테스트에 여러 개의 assert가 존재할 때 이들을 assertAll()로 묶는 것과 묶지 않는 것에는 어떤 차이가 있을까요? 그리고 어떤 것이 더 좋을까요?
4일때와 3일때의 검증 로직을 모두 가지고 있는대요. 이 때는 @ParameterizedTest를 사용하는게 어떨까요?

pine14 replied:
assertAll()로 묶지 않으면 여러 개의 assert 를 실행하는 도중 실패한다면, 그 다음 assert로는 진행하지 않는 것으로 알고 있습니다.
assertAll()을 이용하면 검증이 실패해도 테스트가 끝나지 않고 모든 검증을 완료합니다.

그에 따라 실패한 검증과 성공한 검증을 모두 한눈에 볼 수 있습니다.

한 테스트에는 개념 하나를 테스트하는 것이 좋으므로(클린 코드 166page), 개념 하나를 한 메서드에서 테스트할 때 assert문을 여러 개 사용해야 한다면 assertAll()로 묶는 것이 좋다고 생각합니다.

@ParameterizedTest를 이용해 assert문 두개를 리팩토링 해보았습니다!

euichaan replied:
`@ParameterizedTest(name = "랜덤 숫자 : {0}, 전진 결과 : {1}")
	@MethodSource("randomNumberGenerate")
	@DisplayName("랜덤값이 4 이상이면 전진한다.")
	void move_forward_if_random_number_greater_than_four(int randomNumber, boolean canMove) {
		assertThat(car.shouldMove(randomNumber)).isEqualTo(canMove);
	}

	static Stream<Arguments> randomNumberGenerate() {
		return Stream.of(
			Arguments.of(4, true),
			Arguments.of(3, false)
		);
	}`

pine14 replied:
```java
@ParameterizedTest(name = "랜덤 숫자 : {0}, 전진 결과 : {1}")
	@MethodSource("randomNumberGenerate")
	@DisplayName("랜덤값이 4 이상이면 전진한다.")
	void move_forward_if_random_number_greater_than_four(int randomNumber, boolean canMove) {
		assertThat(car.shouldMove(randomNumber)).isEqualTo(canMove);
	}

	static Stream<Arguments> randomNumberGenerate() {
		return Stream.of(
			Arguments.of(4, true),
			Arguments.of(3, false)
		);
	}
```
다음과 같이 수정해보았습니다.

euichaan replied:
다음과 같이 수정해보았습니다!

skullkim replied:
assertAll()에 대해 의도한 답변을 정확하게 답변해주셨네요😊. @ParameterizedTest를 사용할 때 @MethodSource를 사용해도 문제가 되지는 않아요. 하지만 간단한 파라미터를 사용해야 하는 경우라면 @CsvSource()나 @ValueSource()를 사용하는건 어떨까요?

pine14 replied:
@CsvSource를 이용해서 다음과 같이 처리할 수 있었습니다! 
```java
@ParameterizedTest(name = "랜덤 숫자 : {0}, 전진 결과 : {1}")
	@CsvSource({"4, true", "3, false"})
	@DisplayName("랜덤값이 4 이상이면 전진한다.")
	void move_forward_if_random_number_greater_than_four(int randomNumber, boolean canMove) {
		assertThat(car.shouldMove(randomNumber)).isEqualTo(canMove);
	}
```
그런데 @ValueSource는 어떻게 이용할 수 있을까요..? 4일 때 true, 3일 때 false라는 결과를 assert문 하나로 만들고 싶은데 가능할까요..?

skullkim replied:
@valuesource는 타입이 일정할 때 사용하기 위해 나온 어노테니션이에요. 그래서 @ValuseSource의 문서화 주석을 보면 다음과 같이 설명되서 있어요😃.
```
@ValueSource is an ArgumentsSource which provides access to an array of literal values.
``` 
그래서 이 경우는 @CsvSource를 사용하는 것이 올바른 방법이에요. 혹시 @ValueSource를 써야 하는 이유가 있을까요?

pine14 replied:
`@ValueSource`에 대한 이해가 부족했던 것 같습니다.
이 경우 `@CsvSource`를 사용하는 것이 맞는 방법인 것 같습니다.

skullkim commented on src/test/java/racingcar/util/validator/NameValidatorTest.java:
```diff
@@ -0,0 +1,66 @@
+
+public class NameValidatorTest {
+
+	private Car car = new Car("chan");
```
사용하지 않는 필드는 제거하고 커밋해주세요~

pine14 replied:
네 알겠습니다!👍

skullkim commented on src/test/java/racingcar/util/validator/NameValidatorTest.java:
```diff
@@ -0,0 +1,66 @@
+	@DisplayName("자동차 이름은 Null일 수 없다.")
+	void car_name_should_not_be_null() {
+		assertThatThrownBy(() -> NameValidator.validateCarNames(null))
+			.isInstanceOf(IllegalArgumentException.class);
```
예외에 포함되는 메시지 까지 검사하는게 더 세심한 검사를 할 수 있지 않을까요?

pine14 replied:
AssertJ에 hasMessageContaining이라는 메서드도 있었네요!!

```java
@Test
@DisplayName("자동차 이름은 Null일 수 없다.")
void car_name_should_not_be_null() {
	assertThatThrownBy(() -> NameValidator.validateCarNames(null))
		.isInstanceOf(IllegalArgumentException.class)
		.hasMessageContaining(NULL_CAR_NAME_MESSAGE);
}
```
와 같이 예외에 포함되는 메시지까지 검사해주었습니다.

pine14 replied:
원래는 NameValidatorTest 필드에 전부 상수들을 선언해두고 사용했는데,
NameValidator에 있는 상수들을 사용하고 싶어졌습니다..

```java
private static final String NAME_LENGTH_EXCEPTION_MESSAGE = "자동차의 이름은 1자 이상 5자 이하여야 합니다.";
```
이런 상수들을 public으로 바꾼 후 static import해서 사용해도 문제가 없을까요..?

skullkim replied:
저는 별로 추천하지 않아요~. 저는 테스트 코드가 프로덕션 코드에 어떤 영향도 주어선 안된다고 생각하거든요. 그리고 TDD를 제대로 했다면 테스트를 먼저 작성 했을 테니 에러 메시지에 대한 오탈자 체크도 될거고요.
하지만 사람에 따라 상수 정도는 괜찮다고 생각할 수도 있을거 같아요! 이 부분은 의찬님의 의견도 들어보고 싶네요😄. 의찬님은 테스트 코드를 위해 상수의 접근 제어 정도는 변경해도 된다고 생각하시나요?

pine14 replied:
F.I.R.S.T 의 Timely를 지켰다면 테스트를 먼저 작성했을 테니 테스트 클래스에서 사용할 상수를 먼저 선언했을 것 같습니다!

테스트 케이스에 다음과 같이 선언하고, NameValidator의 상수 접근 제어를 public -> private으로 변경했습니다.
```java
private static final String NAME_LENGTH_EXCEPTION_MESSAGE = "자동차의 이름은 1자 이상 5자 이하여야 합니다.";
private static final String CAR_NAME_NOT_EXIST_MESSAGE = "자동차의 이름이 존재하지 않습니다.";
private static final String DUPLICATE_CAR_NAME_MESSAGE = "자동차의 이름은 중복일 수 없습니다.";
```

skullkim commented on src/test/java/racingcar/util/validator/NameValidatorTest.java:
```diff
@@ -0,0 +1,66 @@
+	@ParameterizedTest(name = "{index}번 자동차 이름 : {0}, 결과 : {1}")
+	@DisplayName("자동차 이름 유효성 검사")
+	@MethodSource("invalidCarNames")
+	void validate_car_name(String carName, String expectedResult) {
```
사용하지 않는 파라미터는 제거하고 커밋해주세요~

pine14 replied:
```java
@ParameterizedTest(name = "{index}번 자동차 이름 : {0}, 결과 : {1}")
	@DisplayName("자동차 이름 유효성 검사")
	@MethodSource("invalidCarNames")
	void validate_car_name(String carName, String expectedResult) {
		assertThatThrownBy(() -> new Car(carName))
			.isInstanceOf(IllegalArgumentException.class);
	}

	static Stream<Arguments> invalidCarNames() {
		return Stream.of(
			Arguments.of("euichan", "이름 5자 초과"),
			Arguments.of("", "1글자 미만"),
			Arguments.of(null, "null 입력"),
			Arguments.of("chan,,dong", "이름이 0자")
		);
	}
```
다음과 같은 테스트인데요, **@ParameterizedTest의** name value로 자동차 이름에 따른 결과를 보고 싶어 Arguments.of에 자동차의 이름과, 예외 조건을 넘겼습니다. 

expectedResult라는 파라미터가 있어야 "이름 5자 초과" 와 같은 예외사항이 나오는데, 제가 제대로 이해하고 있는 것이 맞는지 궁금합니다.

skullkim replied:
Car에서 NameValidator를 사용하고 있잔아요? 그리고 NameValidator에 대한 검증을 다른 테스트에서 했고요. 그러면 Car 생성자 호출을 통한 생성자 검증은 하지 않아도 되지 않을까요?

pine14 replied:
아 그렇네요! 

```java
@ParameterizedTest(name = "{index}번 자동차 이름 : {0}")
@DisplayName("자동차 이름 유효성 검사")
@ValueSource(strings = {"", "dongho", "chan, chan"})
void validate_car_name(String invalidNames) {
	assertThatThrownBy(() -> validateCarNames(List.of(invalidNames)))
		.isInstanceOf(IllegalArgumentException.class);
}
```
다음과 같이 검사하는 것으로 수정했습니다.

skullkim commented on src/test/java/racingcar/util/validator/NumberValidatorTest.java:
```diff
@@ -0,0 +1,25 @@
+
+class NumberValidatorTest {
+
+	private static final String ZERO = "0";
```
사용하지 않은 필드!

pine14 replied:
제거 완료했습니다!

skullkim commented on src/main/java/racingcar/util/validator/NameValidator.java:
```diff
@@ -0,0 +1,55 @@
+	public static void validateCarNames(String carNames) {
+		checkCarNameIsNull(carNames);
+		checkCarNameIsEmpty(carNames);
+		checkCarNameIsDuplicate(carNames);
+	}
+
+	public static void validateCarName(String carName) {
+		checkCarNameIsNull(carName);
+		checkCarNameIsEmpty(carName);
+		checkCarNameLength(carName);
+	}
+
+	private static void checkCarNameIsNull(String carName) {
+		if (carName == null) {
+			throw new IllegalArgumentException(NULL_CAR_NAME_MESSAGE);
+		}
+	}
+
+	private static void checkCarNameIsEmpty(String carName) {
+		if (isBlank(carName)) {
+			throw new IllegalArgumentException(EMPTY_CAR_NAME_MESSAGE);
+		}
+	}
+
+	private static boolean isBlank(String carName) {
+		return carName.isEmpty() || carName.trim().isEmpty();
+	}
+
+	private static void checkCarNameIsDuplicate(String carNames) {
+		List<String> carNameList = Arrays.asList(carNames.split(SEPARATE_BY_COMMA));
+		if (carNameList.size() != carNameList.stream().distinct().count()) {
+			throw new IllegalArgumentException(DUPLICATE_CAR_NAME_MESSAGE);
+		}
+	}
+
+	private static void checkCarNameLength(String carName) {
+		if (carName.length() < MIN_LENGTH_CAR_NAME || carName.length() > MAX_LENGTH_CAR_NAME) {
+			throw new IllegalArgumentException(NAME_LENGTH_EXCEPTION_MESSAGE);
+		}
+	}
```
이 로직을 이렇게 바꾸면 한번의 메서드 호출로 검증을 끝낼 수 있지 않을까요? 이 방법에 대해선 어떤거 같으세요?
```suggestion
public static void validateCarNames(List<String> names) {
		validateTheNumberOfCarNames(names);
		validateDuplicatedCarNames(names);
		validateNameLength(names);
	}
	
	private static void validateTheNumberOfCarNames(List<String> names) {
		if (names == null || isBlank(names)) {
			throw new IllegalArgumentException("자동차 이름이 존재하지 않습니다.");
		}
	}

	private static boolean isBlank(List<String> names) {
		return names.stream()
				.anyMatch(name -> name.isBlank());
	}
	
	private static void validateDuplicatedCarNames(List<String> names) {
		List<String> uniqueNames = names.stream()
				.distinct()
				.collect(Collectors.toList());
		if (names.size() != uniqueNames.size()) {
			throw new IllegalArgumentException(DUPLICATE_CAR_NAME_MESSAGE);
		}
	}
	
	private static void validateNameLength(List<String> names) {
		if (hasInvalidNameLength(names)) {
			throw new IllegalArgumentException(NAME_LENGTH_EXCEPTION_MESSAGE);
		}
	}
	
	private static boolean hasInvalidNameLength(List<String> names) {
		return names.stream()
				.anyMatch(name -> name.length() < MIN_LENGTH_CAR_NAME 
						|| name.length() > MAX_LENGTH_CAR_NAME);
	}
	

```

pine14 replied:
사실 이 부분에 대한 고민이 많았는데
**validateCarNames** 메서드 한 번의 호출로 해결할 수 있어 코드가 깔끔해진 것 같습니다.

추가적으로 실제로 사용하는 코드에서도 한번만 호출을 해주어 코드가 간결해졌습니다.

pine14 replied:
만들어주신 로직으로 코드를 짜다보니 궁금한 점이 생겼는데요..

원래는 Car 의 생성자에서 `validateCarName(name)` 이용해 검증을 진행했습니다.
List<String>을 파라미터로 가지는 `validateCarNames`로 한번에 검증을 진행하니,

Car 생성자에서의 이름 예외처리는 삭제하였습니다.
new Car를 호출하는 부분은 Cars 일급 컬렉션에서 이름을 더해주는 코드 하나입니다.
```java
public void addCarNames(List<String> cars) {
		NameValidator.validateCarNames(cars);
		for (String carName : cars) {
			this.cars.add(new Car(carName));
		}
	}
```
Car를 인스턴스화 할 때 사용하는 이름을 validateCarNames 에서 검증하므로 Car에는 null이 들어갈 수 없다.
따라서 Car는 null-safe 하다고 보는 것이 맞을까요?

skullkim replied:
흠... 이 부분은 자동차 경주 두 번째 PR에서 알려드릴려고 했던 부분과 관련이 있네요. 이 부분에 대해서는 우선 "객체의 책임은 어디까지 일까?"라는 고민을 해보시는게 좋을거 같아요! 그리고 다음 PR에서 알려드린다 했으니 꼭 수정하지는 않으셔도 되요.

pine14 replied:
다른 팩터리 메서드가 생긴다면 검증을 할 수 없기 때문에 검증 부분을 생성자로 두는 게 맞는 것 같습니다!

그렇다고 생성자에 검증 부분을 넣자니, Car 클래스가 이름 검증까지 하는 책임을 가지게 되었습니다.
Name, Position이라는 각각의 클래스로 분리해서 책임을 나눠 보도록 하겠습니다!

**단일 책임 원칙 - 클래스는 오직 하나의 일만 수행해야 생산성이 높아진다. 시스템의 모든 객체는 하나만의 책임을 가져야 한다. 객체가 제공하는 모든 기능은 단 하나의 책임을 수행하는데 집중되야 한다.**

에 따라 Car 클래스는 "전진한다" 라는 책임만을 가지는게 좋을까요?

skullkim commented on src/main/java/racingcar/util/validator/NumberValidator.java:
```diff
@@ -0,0 +1,27 @@
+
+	private static void checkNumberRange(String attemptNumber) {
+		int attemptNumberToInteger = Integer.parseInt(attemptNumber);
+		if (attemptNumberToInteger < 1) {
```
매직넘버~

pine14 replied:
MIN_ATTEMPT_NUMBER 라는 명명된 상수로 교체했습니다.

skullkim reviewed (changes requested):
의찬님 안녕하세요~ 보내주신 PR 잘 봤어요!. 혼자 많이 공부하신게 코드에 다 드러나네요ㅎㅎ. 몇가지 리뷰를 남겼으니 확인해주세요~. 그리고 [클린코드 챕터 4. 주석](https://www.skullkim-dev.com/2022/01/10/4.-%EC%A3%BC%EC%84%9D/)을 보시고 좋은 주석은 무엇일지 한번 고려해 보시면 좋을거 같아요~.

Commit:
refactor : isCustom() 메서드명을 isCustomDelimiter()로 변경

주석을 제거하고 메서드 이름만으로 의도를 더 잘 나타내도록 변경했습니다.

:

Commit:
refactor : 메서드 체이닝 줄바꿈 변경

가독성을 위해 메서드 하나 당 한줄을 차지하게 변경했습니다.

Commit:
refactor : Pattern 성능 개선

pattern을 여러 번 사용하는 경우, 한 번 컴파일 한 후 재사용하는 것이
매번 이 메서드를 호출하는 것보다 효율적입니다.

패턴을 만들고 한 번만 사용할지라도, 나중에는 여러 번 사용할 가능성도 있기 때문에
컴파일 후 재사용하는 방식을 선택하는 것이 훨씬 효율적입니다.

Commit:
refactor : 매직 넘버 제거

매직 넘버를 명명된 상수로 교체했습니다.
숫자는 그 자체로 어떤 의미를 가지지 않습니다.

Commit:
refactor : throws Exception 제거

테스트 메서드에서 예외를 발생시키지 않으므로 throws Exception을 제거했습니다.

Commit:
refactor : 자동차 전진 테스트 리팩터링

랜덤 숫자가 4일때와 3일때의 검증 로직을 모두 가지고 있으므로, ParameterizedTest를 사용해 리팩터링했습니다.

Commit:
refactor : 사용하지 않는 필드 제거

Commit:
refactor : 예외에 포함되는 메시지까지 검사

hasMessageContaining을 이용해서 예외에 포함되는 메시지까지 검사했습니다.

Commit:
refactor : 사용하지 않는 필드 제거

Commit:
refactor : 매직 넘버를 피하기

매직넘버의 사용은 코드의 가독성을 어렵게 만드므로 상수를 사용했습니다.

Commit:
refactor : 자동차 이름 검증 로직을 한번의 메서드 호출로 변경

기존 코드는 두 번에 나눠서 자동차 이름 검증을 진행했지만,
List<String>을 파라미터로 받아 한번에 검증을 수행하도록 했습니다.

Commit:
refactor : 이름 검증 로직에 맞게 테스트 수정

테스트를 수정한 후 hasMessageContaining을 이용해 메시지까지 검증했습니다.

skullkim commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,44 @@
+		return randomNumber >= MOVE_FORWARD_RANDOM_NUMBER_STANDARD;
+	}
+
+	public boolean isWinner(int maxPosition) {
+		if (this.getPosition() == maxPosition) {
+			return true;
+		}
+		return false;
+	}
```
이게 더 깔끔할거 같아요~
```suggestion
	public boolean isWinner(int maxPosition) {
		return this.getPosition() == maxPosition;
	}
```

pine14 replied:
수정 완료했습니다 감사합니다!

return문 한줄로 줄이니 보기 좋네요 ㅎㅎ👍

skullkim commented on src/main/java/racingcar/view/OutputData.java:
```diff
@@ -0,0 +1,44 @@
+	
+	private static StringBuilder printCarsCurrentPosition(Car car) {
+		StringBuilder builder = new StringBuilder();
+		for (int i = 0; i < car.getPosition(); ++i) {
+			builder.append("-");
+		}
```
이건 어떄요?
```suggestion
builder.append("-".repeat(Math.max(0, car.getPosition())));
```

pine14 replied:
몰랐는데 Java11에 도입된 repeat 함수가 있었네요! 

확실히 가독성이 더 좋아진 것 같습니다. 수정 완료했습니다.👍

skullkim reviewed (changes requested):
의찬님~ 리뷰 잘 반영해주셨네요😄. 몇 가지 추가 리뷰와 아직 더 수정이 필요한 부분에 대해 추가 코멘트를 남겼어요! 확인하시고 수정해주세요~. 

그리고 자잘한 컨벤션에 대해서도 신경써 주시면 코드가 더 깔끔해 질거 같아요! 예를 들어서 클래스 내부 첫 줄을 공백으로 통일시킬지 말지 같은 부분이요. 

그리고 클래스 내에서 메서드를 위치하는 방식은 크게 public을 먼저 위치시키고 private을 아래로 놓는 방식과 호출 순으로 메서드를 놓은 방식이 있는대요! 이 부분도 어느 것이 좋을지를 고려한 뒤 선택해 주시면 좋을거 같아요~.
이와 관련된 내용이 클린코드에 나와있는대요~. [5장. 형식 맞추기](https://www.skullkim-dev.com/2022/01/13/5.-%ED%98%95%EC%8B%9D-%EB%A7%9E%EC%B6%94%EA%B8%B0/)를 참고해 고려해주세요~
